### ISSUE TITLE
Cached crate metadata falls out of sync

### ISSUE THREAD

**Issue by djc** [rust-lang/rust-analyzer#17423]
**rust-analyzer version**: rust-analyzer version: 0.3.1995-standalone (c86d6230f 2024-06-10) [/Users/djc/.vscode/extensions/rust-lang.rust-analyzer-0.3.1995-darwin-arm64/server/rust-analyzer]

**rustc version**: 1.79.0 (129f3b996 2024-06-10)

**editor or extension**: VS Code v0.3.1995

**relevant settings**: nothing special

**repository link (if public, optional)**: N/A

In the largish Cargo workspace (~30 crates) at work, I'm seeing behavior where doing something like a `git pull` and then triggering RA to reanalyze will often fail to refresh upstream crates within the workspace. We have a bunch of library small crates and then a few larger binary crates downstream of that, and if API changes happen to my `utils` crate (which my application crate depends on via a `path = "../utils"` dependency) -- like changing the number of function arguments or adding new fields to a type, RA will signal breakage in the downstream users until I explicitly trigger RA to refresh the upstream crate (generally by hitting Cmd + S to save the file).

I forget if RA ever got this right, but if it did it's been a long time since it has.

My expectation would be that, once a refresh is triggered, all files in the workspace could be checked (maybe filtering by mtime) such that the rebuild can happen for all crates that need it.

**Comment by Veykril** [comment]
I'm not quite sure what this is about specifically. Is this about r-a's analysis not seeing the updated signatures or is this about `cargo check` diagnostics being stale? (asking because of " until I explicitly trigger RA to refresh the upstream crate (generally by hitting Cmd + S to save the file)." which can only affect the latter, not the former)

**Comment by djc** [comment]
It is about red squiggles showing up where they shouldn't.

**Comment by davidbarsky** [comment]
mtime isn't sufficient (I believe this is a long-standing bug in Cargo...), but I've found that setting `"rust-analyzer.files.watcher": "server"` helps a bunch with this. VS Code's default file watching behavior isn't particularly reliable.

**Comment by djc** [comment]
Thanks, going to try that. @Veykril any considerations on why that shouldn't be the default?

**Comment by Veykril** [comment]
The LSP spec recommends client file watching hence the default, though for vscode we should probably default to server as vscode's watcher is broken.

Also sorry but I still don't understand the issue here, red squiggles doesn't really answer my question 😅

**Comment by Veykril** [comment]
Actually that doesn't matter I can answer either way, if it's about r-a not recognizing the functions having changed for it's analysis then changing the file watcher should fix it. If it's about cargo check reporting wrong things, then that's the rustc wrapper breaking things which is known and not fixable by us (at best you can disable the wrapper via config)

**Comment by djc** [comment]
It's about r-a not recognizing the functions having changed. IIRC in all of these cases running `cargo check` in the Terminal pane worked correctly.

**Comment by lnicola** [comment]
You should probably whether it actually fixes the problem. But server-side watching had its own share of issues in the past, causing us to disable it (https://github.com/rust-lang/rust-analyzer/issues/1541, linking to https://github.com/notify-rs/notify/issues/208), and we don't know how well it works.

**Comment by djc** [comment]
Yup, it seems I'm still running into issues with it -- it seems to get stuck without making progress, with pretty high CPU usage. Going to have to revert to the client watcher again for now.
